Counter API/Driver updates

nashif opened the pull request:
Consolidated PR against master to catch any issues early.

nordic-krch committed:
drivers: counter: Counter API implementation for nRF Series (TIMER).

Shim for counter API using nrfx_timer driver.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: counter: API unifying counter and rtc drivers.

counter.h and rtc.h API's share almost the same functionality.
They are unified as a new counter.h. rtc.h will be deprecated.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: counter: Counter API implementation for nRF Series (RTC).

Shim for counter API using nrfx_rtc driver.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: counter: Counter API implementation for nRF Series (RTC).

Shim for counter API using nrfx_rtc driver.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: counter: Add tests for counter driver

Add test suite which covers main functionality of counter driver.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

SavinayDharmappa committed:
drivers: counter: Move AON and AONPT driver api to unified interface

patch modifies AON and AONPT driver api to unified interface

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

MaureenHelm committed:
drivers: counter: Introduce mcux rtc shim

Adds a shim layer around the mcux rtc driver to adapt it to the zephyr
counter interface. Portions of this driver are reused from the existing
rtc driver in drivers/rtc/rtc_mcux.c.

The hardware supports a single alarm only and a fixed wrap value.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
soc: nxp_kinetis: Enable the mcux rtc driver on kw41

Enables the mcux rtc shim driver on the kw41 SoC when the counter driver
interface is enabled.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: frdm_kw41z: Document counter driver is supported

Documents that the counter driver is supported in the frdm_kw41z board.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

nordic-krch committed:
drivers: counter: rtc:  Deprecating legacy counter and rtc API

Deprecating counter legacy API and renaming new counter_set_alarm
to coutner_set_ch_alarm. Deprecating rtc API and modifying rtc
API to call new counter API (compatibility layer).

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

MaureenHelm committed:
tests: counter: Adapt test for the mcux rtc driver

The mcux rtc hardware supports a single alarm only and a fixed top
value, so restructure the counter_basic_api test to skip unsupported
features.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
tests: counter: Enable debug logging

Enables debug logging in the counter_basic_api test. Fixes a build error
in the nrf counter drivers when logging is enabled.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

stanislav-poboril committed:
imx: Add IMX EPIT driver for i.MX socs

Add shim driver for i.MX EPIT (Enhanced Periodic Interrupt Timer)
peripheral which can be used for i.MX6SoloX, i.MX7D and other i.MX socs.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
imx: Add EPIT peripheral support for i.MX6SoloX soc

Add EPIT (Enhanced Periodic Interrupt Timer) peripheral support
for i.MX6SoloX soc.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
imx: Enabled EPIT in Udoo Neo Full board

Enabled EPIT (Enhanced Periodic Interrupt Timer) in Cortex-M4 core
of Udoo Neo Full board so the counter API could be used there.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

SavinayDharmappa committed:
drivers: rtc_qmsi: Move rtc driver apis to unified interface

patch modifies rtc apis to unified interface.

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

nordic-krch committed:
drivers: counter: Add user_data to alarm callback

Modify alarm callback to return user_data and channel_id.
Set_alarm and disable_alarm updated accordingly. Renamed
counter_*_ch_alarm to counter_*_channel_alarm. Updated test
and nrf implementations.

Updated doxygen comments.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

erwango committed:
drivers/counter: stm32: Provide STM32 RTC driver using counter API

Provides a STM32 RTC driver using new counter API.
Driver does not support wrap related functions (set, get, ..)

Fixes #11373

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/counter: stm32 rtc: Use dts to get RTC clock information

RTC clock information should be provided by dts files

Fixes #10451

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
samples/drivers: Provides basic sample for counter alarm

Provide basic sample to exercise counter alarm API.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

stanislav-poboril committed:
tests: drivers: counter: Enabled counter tests on UDOO Neo board

Tests i.MX EPIT timer peripheral driver. Since it supports zero number
of channel alarms, some tests had to be adjusted to look at the
number of channel alarms supported first.

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

SavinayDharmappa committed:
soc: x86: Remove RTC config

remove using RTC config

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa committed:
rtc: rename CONFIG_RTC_0_NAME to DT_RTC_0_NAME.

patch renames CONFIG_RTC_0_NAME to DT_RTC_0_NAME.

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

nordic-krch committed:
boards: Add counter capability to nordic boards

Added counter capability to pca10028, pca10040 and pca10056

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
soc: arm: nordic_nrf: Remove duplicates from Kconfig.soc

Removing duplicates in selects for config SOC_NRF52832 and
SOC_NRF52840. Duplicates were unexpected merge artifacts.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

pabigot committed:
tests: drivers: counter: improve comments on reserved instances

RTC1 (not RTC0) is excluded because it is used as the system clock.

Also document that TIMER0 and RTC0 may be rejected when CONFIG_BT_LL_SW
is enabled.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
drivers: counter: nrf: make name configuration consistent

Using `COUNTER_2_NAME` when all other properties are
`COUNTER_TIMER2_foo` is confusing.  Make the names consistent.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
drivers: timer: conflict RTC timer with RTC1 counter

The system timer uses RTC1, but does not implement the counter API with
it.  Instead of auto-enabling the counter API on the system timer make
the two conflict until/unless both APIs are supported by the peripheral.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
drivers: counter: add API to query counter frequency

Add API to get the base counter frequency.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

nashif committed:
tests: remove aon_counter test

This is now covered by counter API.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

pabigot committed:
drivers: counter: nrfx: update to newer HAL API

Upstream trunk moved to a newer Nordic HAL that changed file and
constant names.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nordic-krch committed:
tests: drivers: counter: Fix tests for nrf peripherals

Test was not aligned with changes in the API.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
counter: Fix warning due to deprecation of legacy API

counter_callback_t has been previously prefixed with __deprecated
but it is used in prototype of deprecated function (counter_set_alarm).
It seems that compiler generates the warning even though function is
not used. Removed __deprecated prefix from the typedef, keeping it for
the function only.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: timer: Fix broken dependency for nrf_rtc_timer

nrf_rtc_timer was selecting counter RTC1 instance even though it
is not using counter API at all.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: counter: Rename wrap to top_value

Renamed:
- counter_set_wrap to counter_set_top_value
- counter_get_wrap to counter_get_top_value
- counter_get_max_wrap to counter_get_max_top_value

Updated nRF implementations and counter test.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

pabigot commented:
Is the intent to *not* have a period where topic-counters is left as an integrated PR with a couple days for review?

If this is a DNM advance peek that's fine, but IMO the integrated branch should still go through the normal PR process so we can test cross-platform behavior and resolve conflicts.  It'd be fine to say reviewers can't propose API changes at this stage, but I'm concerned about inconsistency between the implementations that we won't be able to identify until everything's complete and merged to the topic branch.

nashif commented:
> Is the intent to _not_ have a period where topic-counters is left as an integrated PR with a couple days for review?

This to get CI running whenever something is merged into the topic branch, not all tests run there and the topic-branch is not always in sync with master, so this would show conflicts as well. At the end the topic branch will be merged into master using this PR. Work still continues on the topic-branch for now.

dbkinder reviewed (approved):
doc change LGTM

nordic-krch committed:
rtc: Fix call to counter API in adaptation layer

Update rtc.h after renaming counter_set_wrap to counter_set_top_value.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: counter: Fix default top value callback in counter_nrfx_rtc.c

In case of default top value, driver was calling top value
callback only on the first period. This was not expected behavior.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

erwango committed:
boards: nucleo_f746zg: Configure and document counter RTC

Configure RTC on nucleo_f746zg.
It is supported through counter RTC API.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

erwango committed:
tests/drivers: counter: bump cmake minimum version

counter_basic_api test is cmake outdated.
Set it to newest version.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

galak committed:
tests: counter_basic_api: Enable tests for ARM CMSDK drivers

Enable testing of counter (COUNTER_{D}TMR_CMSDK) on mps2_an385 and
enable timers on v2m_beetle (TIMER_{D}TMR_CMSDK).

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: counter: cmsdk: Convert to new DT_<COMPAT>_<INSTANCE> defines

Convert cmsdk driver to use new defines so we can remove the
dts_fixup.h code for it.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
dts: arm_cmsdk_(d)timer: Add label property to (d)timer

Add a label property so we can use that in drivers rather than getting
from Kconfig.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented on drivers/counter/counter_imx_epit.c:
```diff
@@ -0,0 +1,187 @@
+	.set_top_value = imx_epit_set_top_value,
+	.get_pending_int = imx_epit_get_pending_int,
+	.get_top_value = imx_epit_get_top_value,
+	.get_max_relative_alarm = imx_epit_get_max_relative_alarm,
```
@stanislav-poboril wondering why you implemented get_max_relative_alarm if alarm support isn't implemented in EPIT driver?

stanislav-poboril replied:
I think the reason was that it is still valid to call `counter_set_channel_alarm()` when alarm support is not implemented (number of channels is zero) as it will return `-ENOTSUP` . While calling `counter_get_max_relative_alarm()` will try to invoke invalid function pointer and crash. On the other hand, why would anybody do it? The `get_max_relative_alarm` can be removed from EPIT counter driver if needed.

galak reviewed (changes requested):
We need some doc that discusses the theory of the counter api, explains what an alarm is, etc.

nordic-krch committed:
counter: Clarify behavior of counter_set_channel_alarm

Explicitly state that alarm is single shot.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

galak commented on include/counter.h:
```diff
@@ -22,35 +23,190 @@
+ * @param channels	Number of channels that can be used for setting alarm,
+ *			see @ref counter_set_alarm.
+ */
+struct counter_config_info {
```
Somewhere we need a comment about how this has to be the first element in the driver's config struct.

galak committed:
drivers: timer: Move Dual Timer API to unified interface

The CMSDK Dual Timer can be used as a timer or as a counter.
The unified interface proposed in #8340 unifies counter.h and rtc.h to
provide a common interface.

This patch modifies the timer implementation of the dual timer to
make it compliant with the new proposed interface.

Signed-off-by: Vincenzo Frascino <vincenzo.frascino@linaro.org>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: timer: Move Timer API to unified interface

The CMSDK Timer can be used as a timer or as a counter.
The unified interface proposed in #8340 unifies counter.h and rtc.h to
provide a common interface.

This patch modifies the timer implementation of the single timer to
make it compliant with the new proposed interface.

Signed-off-by: Vincenzo Frascino <vincenzo.frascino@linaro.org>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented on tests/drivers/counter/counter_basic_api/src/test_counter.c:
```diff
@@ -0,0 +1,377 @@
+
+static void test_all_instances(counter_test_func_t func)
+{
+	for (int i = 0; i < ARRAY_SIZE(devices); i++) {
```
If ARRAY_SIZE() is 0 we should report an error, since no tests are run.  Right now it looks like things pass.

galak commented on tests/drivers/counter/counter_basic_api/src/test_counter.c:
```diff
@@ -0,0 +1,377 @@
+
+void *exp_user_data = (void *)199;
+
+#ifdef CONFIG_COUNTER_MCUX_RTC
```
Why does MCUX require a special case, this feels like something is lacking in the API.

galak commented on tests/drivers/counter/counter_basic_api/src/test_counter.c:
```diff
@@ -0,0 +1,377 @@
+
+void test_set_top_value_with_alarm(void)
+{
+#ifdef CONFIG_COUNTER_MCUX_RTC
```
Again, this feels like an API issue if we need to special case MCUX_RTC.

galak commented on tests/drivers/counter/counter_basic_api/src/test_counter.c:
```diff
@@ -0,0 +1,377 @@
+
+	alarm_cnt = 0;
+
+	if (counter_get_num_of_channels(dev) < 1U) {
```
If this test is not meant to run, we should look at calling ztest_test_skip()

galak reviewed (changes requested):
There doesn't seem to be any test of the get_top_value api.

erwango committed:
tests: counter: Update to support STM32 RTC counter

Update counter_basic_api test to support STM32 RTC counter.
Add device to list and skipp not supported test cases.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
soc: stm32: Convert CONFIG_RTC leftover to DT_

Convert CONFIG_RTC_0_IRQ_PRI to DT_ namespace.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: Update yaml for boards supporting rtc/counter (1/3)

Few boards supporting RTC were missing rtc as supported rtc feature
in yaml files. Fix this.
Add counter to all boards supporting rtc as RTC IP now support
both rtc and counter API.
Change split in 3 parts in order to lower CI load and get shippable
happy.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: Update yaml for boards supporting rtc/counter (2/3)

Few boards supporting RTC were missing rtc as supported rtc feature
in yaml files. Fix this.
Add counter to all boards supporting rtc as RTC IP now support
both rtc and counter API.
Change split in 3 parts in order to lower CI load and get shippable
happy.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: Update yaml for boards supporting rtc/counter (3/3)

Few boards supporting RTC were missing rtc as supported rtc feature
in yaml files. Fix this.
Add counter to all boards supporting rtc as RTC IP now support
both rtc and counter API.
Change split in 3 parts in order to lower CI load and get shippable
happy.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

nashif committed:
samples: fix alarm sample doc and inclusion

Fix reference for alarm sample and how it is included in the
documentation.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
